# Importing libraries that we will use

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

# Load the CSV file into a pandas DataFrame using the read_csv() function.
Importing data from csv that contain the Bitcoin data.

In [2]:
bitcoin = pd.read_csv('./Documents/UCD python for finance/project/data/Bitcoin - Bitcoin.csv')

In [3]:
print(bitcoin)

              Date       Open*        High         Low     Close**  \
0     feb 28, 2023  $23,521.84  $23,585.38  $23,077.65  $23,147.35   
1     feb 27, 2023  $23,561.45  $23,857.89  $23,205.88  $23,522.87   
2     feb 26, 2023  $23,174.15  $23,654.37  $23,084.22  $23,561.21   
3     feb 25, 2023  $23,200.13  $23,210.21  $22,861.56  $23,175.38   
4     feb 24, 2023  $23,946.01  $24,103.70  $23,007.07  $23,198.13   
...            ...         ...         ...         ...         ...   
3589  may 02, 2013     $116.38     $125.60      $92.28     $105.21   
3590  may 01, 2013     $139.00     $139.89     $107.72     $116.99   
3591  Apr 30, 2013     $144.00     $146.93     $134.05     $139.00   
3592  Apr 29, 2013     $134.44     $147.49     $134.00     $144.54   
3593  Apr 28, 2013     $135.30     $135.98     $132.10     $134.21   

               Volume        Market Cap  
0     $20,535,363,434  $446,846,764,116  
1     $22,660,763,494  $454,075,216,420  
2     $16,644,534,842  $454,795,1

For to be sure that the contents of the DataFrame reflect the meaning of my data, we need to clean and inspect the data.

### We will use the .head() method which displays the first 5 rows

In [6]:
print(bitcoin.head())

           Date       Open*        High         Low     Close**  \
0  feb 28, 2023  $23,521.84  $23,585.38  $23,077.65  $23,147.35   
1  feb 27, 2023  $23,561.45  $23,857.89  $23,205.88  $23,522.87   
2  feb 26, 2023  $23,174.15  $23,654.37  $23,084.22  $23,561.21   
3  feb 25, 2023  $23,200.13  $23,210.21  $22,861.56  $23,175.38   
4  feb 24, 2023  $23,946.01  $24,103.70  $23,007.07  $23,198.13   

            Volume        Market Cap  
0  $20,535,363,434  $446,846,764,116  
1  $22,660,763,494  $454,075,216,420  
2  $16,644,534,842  $454,795,165,823  
3  $16,100,721,565  $447,323,985,683  
4  $26,811,744,928  $447,747,058,644  


# Replace missing values

Some important questions to answer that will help to clean the data are:

Which data type would you expect pandas to assign to each column?
What symbol is used to represent a missing value?

Looking the data we will expect 6 columns type float64 and 1 column type datetime64.

We can see missing values from the first printing at the end of the data, we can print the last 5 rows to check, we will use .tail() function to display the last 5 rows.

In [7]:
print(bitcoin.tail())

              Date    Open*     High      Low  Close** Volume      Market Cap
3589  may 02, 2013  $116.38  $125.60   $92.28  $105.21     $0  $1,168,517,495
3590  may 01, 2013  $139.00  $139.89  $107.72  $116.99     $0  $1,298,954,594
3591  Apr 30, 2013  $144.00  $146.93  $134.05  $139.00     $0  $1,542,813,125
3592  Apr 29, 2013  $134.44  $147.49  $134.00  $144.54     $0  $1,603,768,865
3593  Apr 28, 2013  $135.30  $135.98  $132.10  $134.21     $0  $1,488,566,728


The DataFrame show some values in cero in the column 'volume' at the last rows, but are represented with value 0, in case that we need it in future analysis we will extract the column and see until what date we have data.

# Inspect the data
I will use the .info() to identify dtype mismatches in the DataFrame summary, see the content and the missing values.

In [8]:
bitcoin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3594 entries, 0 to 3593
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Date        3594 non-null   object
 1   Open*       3594 non-null   object
 2   High        3594 non-null   object
 3   Low         3594 non-null   object
 4   Close**     3594 non-null   object
 5   Volume      3594 non-null   object
 6   Market Cap  3594 non-null   object
dtypes: object(7)
memory usage: 196.7+ KB


This Dataframe has 3594 entries, in other words has 3594 rows and 7 columns 
as we can see the dtypes are all object, this is one thing that we need to change, then we will be able to work with the data as we want to manipulate numbers, that will help to get some results in our analysis. 

This DataFrame don't show any missing values, but we will use the na_values parameter to replace any missing values

We will use parse_dates parameter to turn things into real datetime types. parse_dates takes a list of columns, in case you want to parse multiple columns into datetimes.

In [9]:
bitcoin = pd.read_csv('./Documents/UCD python for finance/project/data/Bitcoin - Bitcoin.csv', na_values='n/a', parse_dates =['Date'])

In [10]:
print(bitcoin.head())

        Date       Open*        High         Low     Close**           Volume  \
0 2023-02-28  $23,521.84  $23,585.38  $23,077.65  $23,147.35  $20,535,363,434   
1 2023-02-27  $23,561.45  $23,857.89  $23,205.88  $23,522.87  $22,660,763,494   
2 2023-02-26  $23,174.15  $23,654.37  $23,084.22  $23,561.21  $16,644,534,842   
3 2023-02-25  $23,200.13  $23,210.21  $22,861.56  $23,175.38  $16,100,721,565   
4 2023-02-24  $23,946.01  $24,103.70  $23,007.07  $23,198.13  $26,811,744,928   

         Market Cap  
0  $446,846,764,116  
1  $454,075,216,420  
2  $454,795,165,823  
3  $447,323,985,683  
4  $447,747,058,644  


# Inspecting the data and changing dtypes.

In [11]:
bitcoin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3594 entries, 0 to 3593
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   Date        3594 non-null   datetime64[ns]
 1   Open*       3594 non-null   object        
 2   High        3594 non-null   object        
 3   Low         3594 non-null   object        
 4   Close**     3594 non-null   object        
 5   Volume      3594 non-null   object        
 6   Market Cap  3594 non-null   object        
dtypes: datetime64[ns](1), object(6)
memory usage: 196.7+ KB


We can see that using this parameters, na_values and parse_dates, the column Date has changed to dtype like datetime64 and we are sure that we don't have any missing values

Change the other 6 columns dtypes to float

Convert multiple columns by passing a list of column names to the astype method.

bitcoin[['Open*', 'High','Low','Close**','Volume','Market Cap']] = bitcoin[['Open*', 'High','Low','Close**','Volume','Market Cap']].astype(float)

This wasn't working because there are non numeric symbols like '$' and ','.

# Removing the non-numeric characters from the colums

In [16]:
bitcoin['Open*'] = bitcoin['Open*'].str.replace('[^\d\.]', '', regex=True)

Checking if the non numeric characters are gone from the Open* column 

In [18]:
print(bitcoin.head())

        Date     Open*        High         Low     Close**           Volume  \
0 2023-02-28  23521.84  $23,585.38  $23,077.65  $23,147.35  $20,535,363,434   
1 2023-02-27  23561.45  $23,857.89  $23,205.88  $23,522.87  $22,660,763,494   
2 2023-02-26  23174.15  $23,654.37  $23,084.22  $23,561.21  $16,644,534,842   
3 2023-02-25  23200.13  $23,210.21  $22,861.56  $23,175.38  $16,100,721,565   
4 2023-02-24  23946.01  $24,103.70  $23,007.07  $23,198.13  $26,811,744,928   

         Market Cap  
0  $446,846,764,116  
1  $454,075,216,420  
2  $454,795,165,823  
3  $447,323,985,683  
4  $447,747,058,644  


We will remove the non-numerical characters of the rest of the columns

In [20]:
bitcoin['High'] = bitcoin['High'].str.replace('[^\d\.]', '', regex=True)

In [21]:
bitcoin['Low'] = bitcoin['Low'].str.replace('[^\d\.]', '', regex=True)

In [22]:
bitcoin['Close**'] = bitcoin['Close**'].str.replace('[^\d\.]', '', regex=True)

In [23]:
bitcoin['Volume'] = bitcoin['Volume'].str.replace('[^\d\.]', '', regex=True)

In [24]:
bitcoin['Market Cap'] = bitcoin['Market Cap'].str.replace('[^\d\.]', '', regex=True)

Printing the first 5 rows of the DataFrame, checking if the non-numerical characters are gone in the rest of the columns

In [26]:
print(bitcoin.head())

        Date     Open*      High       Low   Close**       Volume  \
0 2023-02-28  23521.84  23585.38  23077.65  23147.35  20535363434   
1 2023-02-27  23561.45  23857.89  23205.88  23522.87  22660763494   
2 2023-02-26  23174.15  23654.37  23084.22  23561.21  16644534842   
3 2023-02-25  23200.13  23210.21  22861.56  23175.38  16100721565   
4 2023-02-24  23946.01  24103.70  23007.07  23198.13  26811744928   

     Market Cap  
0  446846764116  
1  454075216420  
2  454795165823  
3  447323985683  
4  447747058644  


# Change the dtype from object to float64

In [28]:
bitcoin[['Open*', 'High','Low','Close**','Volume','Market Cap']] = bitcoin[['Open*', 'High','Low','Close**','Volume','Market Cap']].astype(float)

Inspect the data .info() and see if the dtypes changed in all the columns.

In [30]:
bitcoin.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3594 entries, 0 to 3593
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   Date        3594 non-null   datetime64[ns]
 1   Open*       3594 non-null   float64       
 2   High        3594 non-null   float64       
 3   Low         3594 non-null   float64       
 4   Close**     3594 non-null   float64       
 5   Volume      3594 non-null   float64       
 6   Market Cap  3594 non-null   float64       
dtypes: datetime64[ns](1), float64(6)
memory usage: 196.7 KB


As we can see the dtypes of the 6 columns changed to float64 and we have one column with datetime64 that represents the Dates.

This change will help if we want to use the data and we want to compare different numbers to analize the prices sensibility.

# Drop duplicates Dates in the DataFrame.

We use the duplicated() method with the subset parameter set to 'Date' to check for duplicates in that column. The duplicated() method returns a boolean series where True indicates that the row is a duplicated and False indicates that it is not.

In [32]:
# check for duplicates
duplicates = bitcoin.duplicated(subset=['Date'])

# get only the duplicated rows
duplicated_rows = bitcoin[duplicates]

# print out the duplicated row using the 'empty' attribute
if duplicated_rows.empty:
    print("There are no duplicated Dates.")
else:
    print(duplicated_rows)

There are no duplicated Dates.


We use boolean indexing to get only the rows where duplicates is True. This gives us a new dataframe called duplicated_rows that contains only the rows with duplicate Date values.

# Indexing the column 'Date'

In [33]:
bitcoin.set_index('Date', inplace=True)

Use the set_index() method to set the 'Date' column as the index of the DataFrame.

Use the inplace=True parameter to modify the DataFrame in place, no creating a new one.

In [35]:
print(bitcoin.head())

               Open*      High       Low   Close**        Volume    Market Cap
Date                                                                          
2023-02-28  23521.84  23585.38  23077.65  23147.35  2.053536e+10  4.468468e+11
2023-02-27  23561.45  23857.89  23205.88  23522.87  2.266076e+10  4.540752e+11
2023-02-26  23174.15  23654.37  23084.22  23561.21  1.664453e+10  4.547952e+11
2023-02-25  23200.13  23210.21  22861.56  23175.38  1.610072e+10  4.473240e+11
2023-02-24  23946.01  24103.70  23007.07  23198.13  2.681174e+10  4.477471e+11


In [36]:
bitcoin.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 3594 entries, 2023-02-28 to 2013-04-28
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Open*       3594 non-null   float64
 1   High        3594 non-null   float64
 2   Low         3594 non-null   float64
 3   Close**     3594 non-null   float64
 4   Volume      3594 non-null   float64
 5   Market Cap  3594 non-null   float64
dtypes: float64(6)
memory usage: 196.5 KB


The DataFrame has a DatetimeIndex with 3594 entries, contain one row for each day of trading days. 
Starting on 28/02/2023 until 28/04/2013.

There are 6 columns, the first 4 columns summarize the price distribution for the days, the first price, highest price, the lowest price, the last price and the last 2 columns represent the Volume and the Market Cap. 

# Sorting by Date

In [40]:
bitcoin_clean = bitcoin.sort_values(by="Date", ascending=True)

In [41]:
print(bitcoin_clean.head())

             Open*    High     Low  Close**  Volume    Market Cap
Date                                                             
2013-04-28  135.30  135.98  132.10   134.21     0.0  1.488567e+09
2013-04-29  134.44  147.49  134.00   144.54     0.0  1.603769e+09
2013-04-30  144.00  146.93  134.05   139.00     0.0  1.542813e+09
2013-05-01  139.00  139.89  107.72   116.99     0.0  1.298955e+09
2013-05-02  116.38  125.60   92.28   105.21     0.0  1.168517e+09


The sort_values( ) function is used to sort by the values along either axis. In this case we use 2 parameters, by= that indicate the column reference, that we want to sort the data and ascending=True allow to get the sorting in ascending order.